### ISSUE TITLE
nvidia mode not working, nvidia modules already in use

### ISSUE THREAD

**Issue by jazz-g** [Askannz/optimus-manager#147]
Can't switch to Nvidia. Startup is set to Nvidia, but loads Intel. There are no optimus-manager files in /etc/X11/xorg.conf.d/, or any files for that matter. Intel loads Plasma fine.

**System info**
OS: Arch Linux x86_64 
Host: HP Pavilion Gaming Laptop 15-dk0xxx 
Kernel: 5.3.1-arch1-1-ARCH 
DE: Plasma 
CPU: Intel i5-9300H (8) @ 4.100GHz 
GPU: Intel UHD Graphics 630 
GPU: NVIDIA GeForce GTX 1050 3 GB Max-Q 
LM: SDDM
Version: Optimus Manager 1.2.1 (latest stable)

**Logs**
[boot_setup.log](https://github.com/Askannz/optimus-manager/files/3665568/boot_setup.log)
[gpu_setup.log](https://github.com/Askannz/optimus-manager/files/3665569/gpu_setup.log)
[prime_setup.log](https://github.com/Askannz/optimus-manager/files/3665570/prime_setup.log)

**Comment by neurobin** [comment]
same for me

**Comment by TheodoreT** [comment]
same problem here, I had to downgrade to 1.2.0 to have it working again

**Comment by wuzhican** [comment]
how can i downgrade to 1.2.0.
wuzhican@wuzhican-pc
OS: Manjaro 18.1.0 Juhraya
Kernel: x86_64 Linux 5.2.11-1-MANJARO
Uptime: 20m
Packages: 1292
Shell: bash
Resolution: 1920x1080
DE: KDE 5.61.0 / Plasma 5.16.4
WM: KWin
GTK Theme: Breath [GTK2/3]
Icon Theme: maia
Font: Noto Sans Regular
Disk: 100G / 229G (47%)
CPU: Intel Core i7-7700HQ @ 8x 3.8GHz [62.0°C]
GPU: GeForce GTX 1050
 RAM: 2152MiB / 11888MiB

**Comment by TheodoreT** [comment]
@wuzhican for manjaro I guess you need to follow this https://wiki.manjaro.org/index.php?title=Downgrading_packages

install the `downgrade` package and then run `sudo downgrade optimus-manager`

**Comment by Askannz** [comment]
I'm really puzzled here. The error is caused by the nvidia modules being already loaded before optimus-manager can manage them. However, 1) they shouldn't be loaded at all because optimus-manager installs a blacklist 2) The code path up to this point is almost identical between v1.2 and v1.2.1.

@TheodoreT @wuzhican if you're still around, can you confirm the same error appears in your `boot_setup.log` with v1.2.1 ?

```
Cannot setup GPU : kernel setup error : Cannot unload modules ['nvidia_drm', 'nvidia_modeset', 'nvidia'] : Failed to execute 'modprobe -r nvidia_drm nvidia_modeset nvidia' : modprobe: FATAL: Module nvidia is in use.
```

**Comment by TheodoreT** [comment]
@Askannz I confirm that I am getting the same error here.

**Comment by Askannz** [comment]
Is any nvidia module loaded if you boot into console mode with the kernel parameter `systemd.unit=multi-user.target` ?

**Comment by TheodoreT** [comment]
@Askannz both with v1.2.1 and v1.2.0 the only module that shows up after booting in console mode and it is related to nvidia is the `i2c_nvidia_gpu`, nothing else. 

By the way which xorg-server are you using? Are you using the patched version that supports prime offloading or the default one from the official repositories. Here I am using the default one if that makes any difference.

**Comment by pedrovanzella** [comment]
Running as root did not work for me

**Comment by wuzhican** [comment]
i'm facing the same problem.
logs:
[Uploading boot_setup.log…]()
[Uploading gpu_setup.log…]()
[Uploading prime_setup.log…]()
 Optimus Manager (Client) version 1.2.1
and i have modify the /etc/mkinitcpio.conf,
after i delete the modules,
it's still doesn't work

**Comment by wuzhican** [comment]
logs:
[boot_setup.log](https://github.com/Askannz/optimus-manager/files/3760273/boot_setup.log)
[gpu_setup.log](https://github.com/Askannz/optimus-manager/files/3760274/gpu_setup.log)
[prime_setup.log](https://github.com/Askannz/optimus-manager/files/3760275/prime_setup.log)

**Comment by wuzhican** [comment]
我似乎找到了问题所在，
我尝试运行了一下这些命令，
sudo rmmod nvidia
sudo rmmod nvidia-u
我忘了后面的两个字母，大概系统提示我要被引用的模块，我都rmmod了。
然后我尝试switch-mode
大概是成功了，至少日志里面没有异常。
logs:
[boot_setup.log](https://github.com/Askannz/optimus-manager/files/3760344/boot_setup.log)
[gpu_setup.log](https://github.com/Askannz/optimus-manager/files/3760345/gpu_setup.log)
[prime_setup.log](https://github.com/Askannz/optimus-manager/files/3760346/prime_setup.log)

[参考这个链接](https://github.com/Bumblebee-Project/Bumblebee/issues/699)

**Comment by wuzhican** [comment]
it's actually runing
![2019-10-23_13-29](https://user-images.githubusercontent.com/38750781/67360746-45b80500-f599-11e9-8c6d-f53aca852829.png)

**Comment by Sino-Huang** [comment]
what is the use of rmmod btw？ rmmod是删除module还是 重置module呀

**Comment by wuzhican** [comment]
不知道，google出来的，看那个参考链接

**Comment by wuzhican** [comment]
根据这几天的观察，我这边好像是因为invidia_uvm模块没有放在modprobe的移除列表里，然后由于依赖关系导致nvidia无法退出，在root用户下运行modprobe -r nvidia_uvm 后就正常了。

**Comment by wuzhican** [comment]
according to my experience,it's because the nvidia_uvm is activing.so the nvidia module cannot be remove.after i run 'modprobe -r nvidia_uvm' as root .the optimous_manager run normal for me.

**Comment by TheodoreT** [comment]
FYI with the latest version (i.e. v1.2.2) and after the last update on the kernel and the nvidia driver since yesterday switching from intel to nvidia works without issues again. I guess at the end the issue could have been because the nvidia guys was messing around as usual :-1: .

**Comment by Askannz** [comment]
I've also pushed a fix for https://github.com/Askannz/optimus-manager/issues/170, which may also have been the cause of the current issue.

Please test the latest master (optimus-manager-git), and let me know if it solves the issue.

**Comment by pedrovanzella** [comment]
I can confirm it's working with the latest commit

**Comment by TheodoreT** [comment]
@Askannz I am still getting problems. The way I can replicate the problem is the following:

- Boot in intel mode (works without issues)
- Switch for first time to nvidia mode (works without issues)
- Switch back to intel (works without issues)
- Switch back to nvidia (not working returns back to intel, complaining that nvidia modules are still active)

If I check which nvidia modules are still active I can see that `nvidia_uvm` and `nvidia` are still there. Thus, if I remove them manually with `rmmod nvidia_uvm nvidia` and then try to switch to nvidia mode then it works again.

The `gpu_setup.log` gives me the following output:

**(switching from working nvidia to intel)**
```
2019-11-16 03:36:24 PM +0100 ==================== 

Optimus Manager (Setup script) version 1.2.2
Setting up the GPU
Checking status of optimus-manager.service
Cleaning up leftover Xorg conf
Removed /etc/X11/xorg.conf.d/10-optimus-manager.conf
Loading config
Requested mode : intel
Checking for GDM display servers
Available kernel modules : ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset']
Setting up base state
Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia'] (if loaded)
Cannot setup GPU : kernel setup error : Cannot unload modules ['nvidia_drm', 'nvidia_modeset', 'nvidia'] : Failed to execute 'modprobe -r nvidia_drm nvidia_modeset nvidia' : modprobe: FATAL: Module nvidia is in use.
```
**(switching from intel to nvidia, not working since nvidia modules were not unloaded successfully)**
```
2019-11-16 03:37:07 PM +0100 ==================== 

Optimus Manager (Setup script) version 1.2.2
Setting up the GPU
Checking status of optimus-manager.service
Cleaning up leftover Xorg conf
Loading config
Requested mode : nvidia
Checking for GDM display servers
Available kernel modules : ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset']
Setting up base state
Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia'] (if loaded)
Cannot setup GPU : kernel setup error : Cannot unload modules ['nvidia_drm', 'nvidia_modeset', 'nvidia'] : Failed to execute 'modprobe -r nvidia_drm nvidia_modeset nvidia' : modprobe: FATAL: Module nvidia is in use.
```
**(after manually removing remaining nvidia modules, switching back to nvidia from intel work again)**
```
2019-11-16 03:37:47 PM +0100 ==================== 

Optimus Manager (Setup script) version 1.2.2
Setting up the GPU
Checking status of optimus-manager.service
Cleaning up leftover Xorg conf
Loading config
Requested mode : nvidia
Checking for GDM display servers
Available kernel modules : ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset']
Setting up base state
Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia'] (if loaded)
Unloading modules ['nouveau'] (if loaded)
Loading module bbswitch
Setting GPU power to ON via bbswitch
Loading module bbswitch
Setting up Nvidia state
Loading module nvidia
Loading module nvidia_drm
Loaded extra Intel Xorg options (0 lines)
Loaded extra Nvidia Xorg options (8 lines)
Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
```

**Comment by jazz-g** [comment]
Latest commit works fine for me. Should I close this issue?

**Comment by cosmicdan** [comment]
Exact same issue as @ttsesm above, except I can't just manually remove module and repeat - optimus-manager refuses to continue because of the previous error when leaving Intel.

What I have to do instead is:
1) Do `sudo systemctl restart optimus-manager`
2) Do nvidia switch via optimus-manager-qt (for some reason it lets me in this case) and login again
3) Discover that half of my Plasma desktop is missing (some toolbars gone) and systemd is maxing out CPU
4) Log-off via CTRL+ALT+DEL
5) Login again and I've a working NVIDIA mode again
6) Switching back to Intel will need me to repeat the above steps

I'd like to minimize this stuff where possible. Do note I'm new to using multiple GPU's and xorg config in general so I apologize for not recording any relevant logging, but I've a strong hunch it's to do with multiple monitor outputs and prime render. Once during my experimentation (e.g. restarting the om service, manually unloading modules, etc) - I logged out of KDE normally and back in and I was randomly in hybrid mode, according to the QT tray indicator.

Also, I *believe* this only started happening once I had a nice setup in KDE to disable the Intel HDMI output while under NVIDIA (which is using DP). Note I did that under KDE Display Config - NOT via xrandr scripts or xorg - because I thought it least intrusive (doesn't apply until after login) but maybe I've introduced a race condition.

I'll post a reply if I manage to fix my situation neatly, but I'd be interested to know what everyone else's setup is in regards to monitors/multihead and whether they're using Plasma or another DE to manage monitors (vs xrandr or xorg).
